internal/datacoord/segment_reference_manager.go

```diff
@@ -0,0 +1,200 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datacoord
+
+import (
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap"
+
+	etcdkv "github.com/milvus-io/milvus/internal/kv/etcd"
+	"github.com/milvus-io/milvus/internal/log"
+)
+
+type SegmentLock struct {
+	segmentID UniqueID
+	nodeID    UniqueID
+	locKey    string
+}
+
+type SegmentReference struct {
+	etcdKV *etcdkv.EtcdKV
+
+	segmentsLock map[UniqueID][]*SegmentLock
+	lock         sync.RWMutex
+}
+
+func parseLockKey(key string) (segID UniqueID, nodeID UniqueID, err error) {
+	ss := strings.Split(key, "/")
+	// segment lock key consists of at least "/segmentRefer/nodeID/segID"
+	if len(ss) < 4 {
+		return -1, -1, fmt.Errorf("segment lock key is invalid with %s", key)
+	}
+	segID, err = strconv.ParseInt(ss[len(ss)-1], 10, 64)
+	if err != nil {
+		return 0, 0, err
+	}
+	nodeID, err = strconv.ParseInt(ss[len(ss)-2], 10, 64)
+	return segID, nodeID, err
+}
+
+func NewSegmentReference(etcdKV *etcdkv.EtcdKV) (*SegmentReference, error) {
+	segReferManager := &SegmentReference{
+		etcdKV:       etcdKV,
+		segmentsLock: make(map[UniqueID][]*SegmentLock),
+	}
+	keys, _, err := segReferManager.etcdKV.LoadWithPrefix(segmentReferPrefix)
+	if err != nil {
+		log.Error("load segments lock from etcd failed", zap.Error(err))
+		return nil, err
+	}
+
+	for _, key := range keys {
+		segID, nodeID, err := parseLockKey(key)
+		if err != nil {
+			log.Error("parse segment lock key failed", zap.Int64("segmentID (maybe wrong)", segID), zap.String("lock key", key),
```

Comment:
Same as above. Maybe we don't need printing the (possible wrong) segment ID here. It should be straightforward to figure out what went wrong by printing out and checking `key`.